Port to TelepathyGLib and Python3

Saumya-Mishra9129 opened the pull request:
Tested works fine. @quozl @chimosky @JuiP @srevinsaju Please review.

yell0wfl4sh committed:
first port to gtk3

yell0wfl4sh committed:
update file to gtk3

yell0wfl4sh committed:
remove unused toolkit files

yell0wfl4sh committed:
update on port

yell0wfl4sh committed:
port to cairo

yell0wfl4sh committed:
remove bugs porting to cairo

yell0wfl4sh committed:
port drawable to set_source_surface

yell0wfl4sh committed:
add missing drawgc.fill()

yell0wfl4sh committed:
fix keyframe drawing

yell0wfl4sh committed:
fix color warnings

yell0wfl4sh committed:
make lessons in toolbar visible

yell0wfl4sh committed:
Remove extra SeparatorToolItem

quozl committed:
Correctly call the parent class initialiser

quozl committed:
Fix logger name

quozl committed:
Remove debug code

quozl committed:
Fix missing slider icons

Call new_from_data during object construction.

yell0wfl4sh committed:
fix color invert issue

Saumya-Mishra9129 committed:
Port to TelepathyGLib

Saumya-Mishra9129 committed:
AttributeError: 'flipsticksActivity' object has no attribute '_shared_activity'

Saumya-Mishra9129 committed:
Change exec to sugar-activity3

Saumya-Mishra9129 committed:
Port to Python3 : Run 2to3

Saumya-Mishra9129 committed:
Remove trailing semicolons

Saumya-Mishra9129 committed:
Port to Python3 : change from file to open

Saumya-Mishra9129 committed:
Change from gobject_service to gi_service

Saumya-Mishra9129 committed:
TypeError: Item 0: Must be number, not str

Saumya-Mishra9129 committed:
fix line width for figure below

quozl commented on activity.py:
```diff
@@ -44,37 +51,40 @@ def __init__(self, handle):
+        activity_button = ActivityToolbarButton(self)
+        toolbar_box.toolbar.insert(activity_button, 0)
+        activity_button.show()
+        self.toolbar_box = toolbar_box
```
What is your purpose for this line?

Saumya-Mishra9129 replied:
No purpose, It was added by Yash when he ported. I forgot to see. It is of no use .

quozl commented on model.py:
```diff
@@ -202,9 +202,9 @@ def screen_shot(pixbuf):
+    pixbuf.savev(filepath,'png')
 
-    from sugar.datastore import datastore
+    from sugar3.datastore import datastore
```
Shouldn't this import be at top of file?

Saumya-Mishra9129 replied:
Thanks I will do

quozl commented on montage.py:
```diff
@@ -121,13 +121,10 @@ def exportframe(self):
+        firstpixindex = list(self.frames.keys())[0]
+        window = Gdk.get_default_root_window()
+        x,y, width , height = window.get_geometry()
+        pixbuf = Gdk.pixbuf_get_from_window(window, x, y, width, height)
```
You've ported from GTK 2 to GTK 3 and reimplemented a screenshot feature.  Activity.get_preview does this now; can it be used instead?

Saumya-Mishra9129 replied:
Thanks I will try with get_preview, hope that will work.

Saumya-Mishra9129 replied:
@quozl I wonder why we need to give screenshots feature in this activity , and also a button in toolbar for the same. We have already option in toolkit to take screenshot with keyboard keys. Why it's necessary?

quozl replied:
Does it offer more than the Sugar feature?  Was the feature added to the activity before it was added to Sugar?  The feature in Sugar is whole display including toolbar.  Does this feature include toolbar?

Saumya-Mishra9129 replied:
I am not sure about displaying toolbar as I couldn't test it. However the way I implemented here, it includes toolbar. The feature was added to activity in 2009. As far as I can see in history of toolkit first commit was made in 2011, and it had get_preview feature.

quozl replied:
Sugar had the global key screenshot feature added in 2009.  See https://github.com/sugarlabs/sugar/blame/master/extensions/globalkey/screenshot.py

Sugar had the frame display icon screenshot feature added six years ago.  See https://github.com/sugarlabs/sugar/blame/master/extensions/deviceicon/display.py#L180

get_preview was mentioned because the code in this activity is similar to get_preview.  I'm not surprised you didn't find when get_preview was added.  Toolkit has a second repository, for the GTK 2 source.  In that, https://github.com/sugarlabs/sugar-toolkit/blame/master/src/sugar/activity/activity.py#L663 shows get_preview was made available (by public API) in 2009.  The activity would have been written for the previous version of Sugar at the time.  The education department of the country that focused on this activity delayed upgrading Sugar on their laptops.

So to answer my questions;

> Does it offer more than the Sugar feature?

Yes, it takes a snapshot of the animation window, not the canvas, not the display.

> Was the feature added to the activity before it was added to Sugar?

Yes.

> The feature in Sugar is whole display including toolbar. Does this feature include toolbar?

No.

Saumya-Mishra9129 replied:
Thanks @quozl . Maybe 1dee462 is a good fix for this activity. I have tested it looks like this , it is opened in ImageViewer.

![Screenshot from 2020-08-07 11-57-01](https://user-images.githubusercontent.com/45252461/89616304-cb493f00-d8a5-11ea-9aeb-878c2997fec6.png)

quozl replied:
The grey box top right is surprising.  Is that the toolbar button palette?

Saumya-Mishra9129 replied:
Yes it is.

Saumya-Mishra9129 replied:
>The grey box top right is surprising. Is that the toolbar button palette?

It can be seen only with smaller screens. Because whole screen becomes too congested.

quozl commented on toolkit/activity.py:
```diff
@@ -146,18 +150,18 @@ def get_cursor(self):
         if self._cursor != cursor:
             self._cursor = cursor
-            self.window.set_cursor(self._cursor)
+            self.get_window().set_cursor(self._cursor)
```
GTK 3 does not make cursor changes immediately.  See Activity.busy for how to flush this.

Saumya-Mishra9129 replied:
Activity.busy where it can be found?

quozl replied:
See [API](https://developer.sugarlabs.org/sugar3/sugar3.activity.activity.html) and search for busy.

See [Activity.busy](https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/activity/activity.py#L1402).

It is the [call to _Gdk.Flush_](https://github.com/sugarlabs/sugar-toolkit-gtk3/blob/master/src/sugar3/activity/activity.py#L1435) needed in recent GTK releases, unless the application uses a spinner or progress bar.

Saumya-Mishra9129 replied:
Tried it in 318b019.

quozl commented on toolkit/activity.py:
```diff
@@ -206,7 +210,7 @@ def read_file(self, filepath):
             self.__state = _PRE_INSTANCE
         elif self.__state == _PRE_INSTANCE:
-            self.__instance();
+            self.__instance()
```
Agreed.  I liked seeing evidence of C programmers.  :grinning:

Saumya-Mishra9129 replied:
:)

chimosky commented:
Tested, reviewed.

On sharing, the user instance of the sharer is no longer seen in the neighbourhood of the joiner and this error is shown in `shell.log`
```log
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1641: Warning: g_value_transform: assertion 'G_IS_VALUE (src_value)' failed
  return _Gtk_main(*args, **kwargs)
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1641: Warning: unable to set property 'buddy' of type 'PyObject' from value of type '(null)'
  return _Gtk_main(*args, **kwargs)
Traceback (most recent call last):
  File "/usr/lib/python3.7/dist-packages/jarabe/model/neighborhood.py", line 1035, in __activity_updated_cb
    bundle = registry.get_bundle(properties['type'])
KeyError: 'type'
1596239678.786034 ERROR root: set_active() failed: org.freedesktop.DBus.Error.NoReply: Message recipient disconnected from message bus without replying
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1641: Warning: g_value_transform: assertion 'G_IS_VALUE (src_value)' failed
  return _Gtk_main(*args, **kwargs)
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1641: Warning: unable to set property 'buddy' of type 'PyObject' from value of type '(null)'
  return _Gtk_main(*args, **kwargs)
1596239683.186974 ERROR root: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.16 was not provided by any .service files
Traceback (most recent call last):
  File "/usr/lib/python3.7/dist-packages/jarabe/model/filetransfer.py", line 345, in file_transfer_available
    CONNECTION_INTERFACE_REQUESTS)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 145, in __call__
    **keywords)
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.16 was not provided by any .service files
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1641: Warning: g_value_transform: assertion 'G_IS_VALUE (src_value)' failed
  return _Gtk_main(*args, **kwargs)
/usr/lib/python3/dist-packages/gi/overrides/Gtk.py:1641: Warning: unable to set property 'buddy' of type 'PyObject' from value of type '(null)'
  return _Gtk_main(*args, **kwargs)
1596239683.444698 ERROR root: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.16 was not provided by any .service files
Traceback (most recent call last):
  File "/usr/lib/python3.7/dist-packages/jarabe/model/filetransfer.py", line 345, in file_transfer_available
    CONNECTION_INTERFACE_REQUESTS)
  File "/usr/lib/python3/dist-packages/dbus/proxies.py", line 145, in __call__
    **keywords)
  File "/usr/lib/python3/dist-packages/dbus/connection.py", line 651, in call_blocking
    message, timeout)
dbus.exceptions.DBusException: org.freedesktop.DBus.Error.ServiceUnknown: The name :1.16 was not provided by any .service files
```

quozl commented:
That's pretty much https://github.com/sugarlabs/sugar/issues/840 and needs fixing in Telepathy.

chimosky commented:
@quozl said
> That's pretty much [sugarlabs/sugar#840](https://github.com/sugarlabs/sugar/issues/840) and needs fixing in Telepathy.

Yeah I think they're related but in this case the joiner doesn't see the sharer in their neighbourhood once the activity is shared, the joiner doesn't join the activity as they're not able to. 
I also noticed after a while the sharer comes back on the joiner's neighbourhood.

Saumya-Mishra9129 commented:
> Yeah I think they're related but in this case the joiner doesn't see the sharer in their neighbourhood once the activity is shared, the joiner doesn't join the activity as they're not able to.
I also noticed after a while the sharer comes back on the joiner's neighbourhood.

Thanks I have also seen this behaviour earlier, we had also noticed this behaviour in https://github.com/sugarlabs/maze-activity/pull/31 , But surprisingly I saw it working. collaboration I tested again, it's working as expected. Can It be something related to your system?? I don't know but maybe it can.

> That's pretty much sugarlabs/sugar#840 and needs fixing in Telepathy.

I think the error above mentioned by @chimosky  is related to sugar not to this activity.

Saumya-Mishra9129 committed:
Remove unnecessary line added in https://github.com/sugarlabs/flipsticks/pull/6/commits/c221185ec80c41a35d31c94e773f745c5f99c6cb

Saumya-Mishra9129 committed:
Move import to the top of the file

Saumya-Mishra9129 committed:
Add Gdk.flush to fix cursor changes

GTK3 does not make cursor changes immediately. Gdk.flush() flushes the output buffers of all display connections and wait until all requests have been processed.

Saumya-Mishra9129 committed:
Port to Gtk3 : Fix Snapshot

Saumya-Mishra9129 commented:
@quozl Please review 1dee462. This feature don't display whole toolbar.

quozl commented:
1dee462 reviewed.

Saumya-Mishra9129 commented:
> 1dee462 reviewed.

Thanks @quozl . Can we merge this? So that we can release a new version with Gtk3 and Python3 port.?

quozl commented:
Sure, the outstanding behaviours can be raised as issues.
